lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
```

Comment:
I did this because we now have not yet done the dry run (which validates the template) at the time we instantiate these classes. So although the template would not be valid if this field were missing, it is not guaranteed to be valid at this point.